### ISSUE TITLE
After install offer to reload VS with current solution

### ISSUE THREAD

**Issue by jcansdale** [jcansdale/DogfoodVsix#2]
Chances are the user will want to get back to where they were.

After installing a .vsix, we could offer to relaunch VS with the current solution.

https://github.com/justcla/VSPackageInstaller/blob/master/VSPackageInstaller/VSPackageInstaller/PackageInstaller/Helpers/VsHelpers.cs#L46-L56

**Comment by jcansdale** [comment]
I'm now using `IVsShell4.Restart`, but it doesn't look like specifying arguments is an option.

https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.shell.interop.ivsshell4.restart?view=visualstudiosdk-2017

**Comment by ironcev** [comment]
As far as I can see, the current solution gets properly reloaded after VS restart. This is BDW exactly the behavior I would expect, without any additional questions. It should be the default behavior just as it is now.

I wanted to ask is this issue actually solved and should be closed, or there is still something to do?

**Comment by jcansdale** [comment]
Hey 👋 

At the moment I'm simply calling `IVsShell4.Restart` which restarts Visual Studio, but not with the active solution. This is better than asking the user to restart and I've found it easy enough to fund the most recent solution (e.g on the Start Page / Recent Projects list).

I'm hesitant to make this the default, because sometimes I'll load an extensibility project, execute Dogfood and then load a different solution. Loading the original solution could be a pain in this scenario, especially for larger solutions!

I'm interested in your thoughts though. Let me know how is would fit with your workflows. 😄

**Comment by ironcev** [comment]
Ok, I think I've connected all the dots now. `IVsShell4.Restart()` restarts VS and automatically opens the solution that was active before the restart. There is no additional argument that could specify to open the last active solution or not. That's why it always reloads the last solution in my case, which is the behavior I expected.

But after giving it a second thought, I agree with you that having this as a default could be contra-productive in some cases, such as the case you described.

Ideally, the "Would you like to restart Visual Studio now?" dialog could have an additional checkbox like:
[ ] Reload current solution after restart

What you're saying is that there is no VS API that could help us here, or at least you are not aware of it.

I thought of a simple (and ugly of course) workaround - to close the current solution before restart in case the user does not want to reload it after restart. In that case after restart no solution will be loaded and the user can pick up one from the Recent Projects list.

But I've simulated that right now and it didn't work?!

Steps to reproduce:
- Open some solution.
- Install some VSIX via Dogfood e.g. from VsixGallery.
- Choose "Cancel" when asked to restart.
- Close the solution.
- Install the same VSIX again.
- This time choose "Ok" when asked to restart.

VS will restart and still open the solution that was opened before being closed before the restart?! Strange.

Anyhow, it's time to go to sleep :-)

Bottom line: the user should be able to choose to automatically reopen the current solution after restart or not. It doesn't work that way at the moment ➡️ issue is still open :-)

**Comment by jcansdale** [comment]
> That's why it always reloads the last solution in my case, which is the behavior I expected.

I wondered why it wasn't working this way for me, but I've just noticed this setting:

![image](https://user-images.githubusercontent.com/11719160/49050109-ba43c180-f1d9-11e8-91e6-3243f44244bf.png)

I'm guessing you have that enabled?

**Comment by ironcev** [comment]
No, "Show Start Page" is my default option. And VS still reloads the last solution after restart ;-)

However, the "Steps to reproduce" from yesterday are not reproducible anymore (VS opens a solution after restart that was not open before restarting). The only difference from today's and yesterday's VS on my machine is the fact that I uninstalled one extension that was failing to load.

So the exact behavior of the `IVsShell4.Restart()` remains a bit of a mystery. The official documentation does not say anything about reloading a solution.

If you want I can dig deeper into this issue next week (isolated example, trying on several machines, StackOverflow?). Maybe it's not that crucial at the moment but I'm curious why the difference. Besides, indeterministic behavior is not the best thing to have in our line of business :-)

**Comment by jcansdale** [comment]
> Maybe it's not that crucial at the moment but I'm curious why the difference. Besides, indeterministic behavior is not the best thing to have in our line of business :-)

I think you're right. Not that crucial, but I'm curious WTH is going on! 😉

**Comment by jcansdale** [comment]
Closing because I think simply reloading VS is good enough.
